<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Marlee Labroo, Christine Nyaga, Lukas Mueller" />


<title>Genomic Prediction of Cross Performance with gpcp</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Genomic Prediction of Cross Performance
with gpcp</h1>
<h4 class="author">Marlee Labroo, Christine Nyaga, Lukas Mueller</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette demonstrates how to use the <code>gpcp</code> package
to perform genomic prediction of cross performance using genotype and
phenotype data. This method processes data in several steps, including
loading the necessary software, converting genotype data, processing
phenotype data, fitting mixed models, and predicting cross performance
based on weighted marker effects.</p>
<p>The package is particularly useful for users working with polyploid
species, and it integrates with the <code>sommer</code>,
<code>AGHmatrix</code>, and <code>snpStats</code> packages for efficient
model fitting and genomic analysis.</p>
</div>
<div id="installing-the-gpcp-package" class="section level1">
<h1>Installing the gpcp Package</h1>
<p>If you haven’t installed the <code>gpcp</code> package yet, you can
do so by following these steps:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Install devtools if you don&#39;t have it</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># Install BiocManager in order to install VariantAnnotatiion and snpStats</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#Install VariantAnnotation and snpStats</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;VariantAnnotation&quot;</span>)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;snpStats&quot;</span>)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co"># Install gpcp from your local repository or GitHub</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;cmn92/gpcp&quot;</span>)</span></code></pre></div>
</div>
<div id="getting-started" class="section level1">
<h1>Getting Started</h1>
<p>The main function in this package is <code>runGPCP()</code>, which
predicts the performance of genomic crosses. To run this function,
you’ll need two main input files: 1. A phenotype file, which is
typically a CSV file containing the phenotypic data. 2. A genotype file,
which can be in VCF or HapMap format.</p>
</div>
<div id="example-workflow" class="section level1">
<h1>Example Workflow</h1>
<p>Let’s walk through a simple example to predict cross performance
using the provided phenotype and genotype data.</p>
<div id="step-1-load-the-required-data" class="section level2">
<h2>Step 1: Load the Required Data</h2>
<p>Before running <code>runGPCP</code>, load the phenotype data from a
CSV file and specify the genotype file path.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Load phenotype data</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>phenotypeFile <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/gpcp/data/phenotypeFile.csv&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Specify the genotype file path (VCF or HapMap format)</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>genotypeFile <span class="ot">&lt;-</span> <span class="st">&quot;~/gpcp/data/genotypeFile_Chr9and11.vcf&quot;</span></span></code></pre></div>
</div>
<div id="step-2-define-the-necessary-inputs" class="section level2">
<h2>Step 2: Define the Necessary Inputs</h2>
<p>You will need to specify several inputs such as the genotypes column,
traits to predict, and other variables such as weights, fixed effects,
and ploidy.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Define inputs</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>genotypes <span class="ot">&lt;-</span> <span class="st">&quot;Accession&quot;</span>  <span class="co"># Column name for genotype IDs in phenotype data</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>traits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;YIELD&quot;</span>, <span class="st">&quot;DMC&quot;</span>)  <span class="co"># Traits to predict</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>)  <span class="co"># Weights for each trait</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>userFixed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;LOC&quot;</span>, <span class="st">&quot;REP&quot;</span>)  <span class="co"># Fixed effects</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>Ploidy <span class="ot">&lt;-</span> <span class="dv">2</span>  <span class="co"># Ploidy level</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>NCrosses <span class="ot">&lt;-</span> <span class="dv">150</span>  <span class="co"># Number of crosses to predict</span></span></code></pre></div>
</div>
<div id="step-3-run-the-genomic-prediction" class="section level2">
<h2>Step 3: Run the Genomic Prediction</h2>
<p>Now that we have the necessary inputs, we can run the
<code>runGPCP()</code> function to predict cross performance.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Run genomic prediction of cross performance</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>finalcrosses <span class="ot">&lt;-</span> <span class="fu">runGPCP</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>    <span class="at">phenotypeFile =</span> phenotypeFile,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    <span class="at">genotypeFile =</span> genotypeFile,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>    <span class="at">genotypes =</span> genotypes,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    <span class="at">traits =</span> <span class="fu">paste</span>(traits, <span class="at">collapse =</span> <span class="st">&quot;,&quot;</span>),</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>    <span class="at">weights =</span> weights,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>    <span class="at">userFixed =</span> <span class="fu">paste</span>(userFixed, <span class="at">collapse =</span> <span class="st">&quot;,&quot;</span>),</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>    <span class="at">Ploidy =</span> Ploidy,</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>    <span class="at">NCrosses =</span> NCrosses</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="step-4-view-the-results" class="section level2">
<h2>Step 4: View the Results</h2>
<p>The output of the <code>runGPCP()</code> function is a data frame
that contains the predicted cross performance. You can view the top
predicted crosses like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># View the predicted crosses</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">head</span>(finalcrosses)</span></code></pre></div>
<p>The resulting data frame contains the following columns: -
<code>Parent1</code>: The first parent of the cross. -
<code>Parent2</code>: The second parent of the cross. -
<code>CrossPredictedMerit</code>: The predicted merit of the cross. -
<code>P1Sex</code> and <code>P2Sex</code>: Optional. If sex information
is provided, the sexes of the parents are included.</p>
</div>
</div>
<div id="details-of-the-process" class="section level1">
<h1>Details of the Process</h1>
<p>The <code>runGPCP()</code> function performs the following steps
internally: 1. <strong>Read the genotype and phenotype data</strong>:
The genotype file is converted into a matrix of allele counts, and the
phenotype data is standardized. 2. <strong>Fit mixed models</strong>:
The <code>sommer</code> package is used to fit mixed models based on
user-defined fixed and random effects. 3. <strong>Predict cross
performance</strong>: Marker effects are calculated and weighted to
predict the performance of crosses, and the best crosses are
identified.</p>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>The methodology behind the <code>gpcp</code> package is based on the
following references: - Xiang, J., et al. (2016). “Mixed Model Methods
for Genomic Prediction.” <em>Nature Genetics</em>. - Batista, L., et
al. (2021). “Genetic Prediction and Relationship Matrices.”
<em>Theoretical and Applied Genetics</em>.</p>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>The <code>gpcp</code> package provides a flexible and efficient
framework for predicting genomic cross performance in both diploid and
polyploid species. With its ability to handle multiple traits, fixed
effects, and random effects, this package is ideal for breeders and
geneticists looking to maximize cross potential using genomic data.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
